### PR TITLE
docs(testing): translate testing-environments

### PR DESCRIPTION
**Pull request by neighborhood999**

### CODE REVIEW

File: content/docs/testing-environments.md [hunk 1]
```diff
@@ -1,58 +1,58 @@
...
+有時候你可能不想要 mock timer。例如，或許你想要測試一個 animation，或是與一個對時間較敏感的 endpoint 交互（像是 API 的 rate limiter）。函式庫的 timer mock 讓你啟用和關閉它們在每個測試的基礎上，所以你可以明確的選擇這些測試的執行方式。
 
-### End-to-end tests {#end-to-end-tests-aka-e2e-tests}
+### End-to-end 測試 {#end-to-end-tests-aka-e2e-tests}
 
-End-to-end tests are useful for testing longer workflows, especially when they're critical to your business (such as payments or signups). For these tests, you'd probably want to test how a real browser renders the whole app, fetches data from the real API endpoints, uses sessions and cookies, navigates between different links. You might also likely want to make assertions not just on the DOM state, but on the backing data as well (e.g. to verify whether the updates have been persisted to the database).
+End-to-end 測試對於測試更長的 workflow 非常有用，特別是當它們對於你的商業邏輯至關重要時（像是付款或是註冊）。對於這些測試，你可能需要測試真實瀏覽器是如何 render 整個應用程式，從真實 API endpoint 來 fetch 資料，使用  session 和 cookie，在不同的連結之間導航。你可能不只想要 assert 在 DOM 上的 state，而且還要備份資料的 assert（例如：驗證更新是否已經被儲存到資料庫）。
```
Comment (bugtender) [inline content/docs/testing-environments.md]:
```suggestion
End-to-end 測試對於測試更長的 workflow 非常有用，特別是當它們對於你的商業邏輯至關重要時（像是付款或是註冊）。對於這些測試，你可能需要測試真實瀏覽器是如何 render 整個應用程式，從真實 API endpoint 來 fetch 資料，使用 session 和 cookie，在不同的連結之間導航。你可能不只想要 assert 在 DOM 上的 state，而且還要備份資料的 assert（例如：驗證更新是否已經被儲存到資料庫）。
```

File: content/docs/testing-environments.md [hunk 2]
```diff
@@ -1,58 +1,58 @@
...
+Component 可能使用基於像是 `setTimeout`、`setInterval` 或是 `Date.now` 的時間函式。在測試環境中，可以透過 mock 這些函式來做替換，對於你想要手動「提前」時間非常有用。這可以確保你的測試很快地執行！依賴於 timer 的測試仍然可以按順序 resolve，但速度更快[<small>（範例）</small>](/docs/testing-recipes.html#timers)。大部分的測試框架，包含 [Jest](https://jestjs.io/docs/en/timer-mocks)、[sinon](https://sinonjs.org/releases/v7.3.2/fake-timers/) 以及 [lolex](https://github.com/sinonjs/lolex) 讓你可以在測試中 mock timer。
 
-Sometimes, you may not want to mock timers. For example, maybe you're testing an animation, or interacting with an endpoint that's sensitive to timing (like an API rate limiter). Libraries with timer mocks let you enable and disable them on a per test/suite basis, so you can explicitly choose how these tests would run.
+有時候你可能不想要 mock timer。例如，或許你想要測試一個 animation，或是與一個對時間較敏感的 endpoint 交互（像是 API 的 rate limiter）。函式庫的 timer mock 讓你啟用和關閉它們在每個測試的基礎上，所以你可以明確的選擇這些測試的執行方式。
```
Comment (bugtender) [inline content/docs/testing-environments.md]:
```suggestion
有時候你可能不想要 mock timer。例如，或許你想要測試一個 animation，或是與一個對時間較敏感的 endpoint 交互（像是 API 的 rate limiter）。函式庫的 timer mock 讓你在每個測試的基礎上啟用和關閉它們，所以你可以明確的選擇這些測試的執行方式。
```

File: content/docs/testing-environments.md [hunk 3]
```diff
@@ -1,58 +1,58 @@
...
+當撰寫測試時，我們會想要 mock 在我們測試環境中沒有等效的程式碼部分（例如：在 Node.js 內判斷 `navigator.onLine` 狀態）。測試也可以監視（spy）某些函式，並觀察其他部分的測試如何與它交互。可以使用易於測試的版本，對於選擇性地來 mock 這些函式很有用。
 
-This is especially useful for data fetching. It is usually preferable to use "fake" data for tests to avoid the slowness and flakiness due to fetching from real API endpoints [<small>(example)</small>](/docs/testing-recipes.html#data-fetching). This helps make the tests predictable. Libraries like [Jest](https://jestjs.io/) and [sinon](https://sinonjs.org/), among others, support mocked functions. For end-to-end tests, mocking network can be more difficult, but you might also want to test the real API endpoints in them anyway.
+這特別是對於資料的 fetch 非常有用。通常最好是使用「假」資料來測試，避免從真實 API endpoint 取得資料，造成緩慢和脆弱的測試[<small>（範例）</small>](/docs/testing-recipes.html#data-fetching)。這可以幫助測試是可預期的。像是 [Jest](https://jestjs.io/) 和 [sinon](https://sinonjs.org/) 的函式庫，都支援 mock 函式。對於 end-to-end 測試來說，mock network 可能會更困難，但你仍然可能也想要在測試中測試真實的 API endpoint。
```
Comment (bugtender) [inline content/docs/testing-environments.md]:
```suggestion
這特別是對於資料的 fetch 非常有用。通常最好是使用「假」資料來測試，避免從真實 API endpoint 取得資料，造成緩慢和脆弱的測試[<small>（範例）</small>](/docs/testing-recipes.html#data-fetching)。這增進了測試的可預期性。像是 [Jest](https://jestjs.io/) 和 [sinon](https://sinonjs.org/) 的函式庫，都支援 mock 函式。對於 end-to-end 測試來說，mock network 可能會更困難，但你仍然可能也想要在測試中測試真實的 API endpoint。
```

File: content/docs/testing-environments.md (hunk 2 above)
Comment (neighborhood999) [reply content/docs/testing-environments.md]: Thanks, I updated.

**Pull request merged**